sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -1112,39 +1072,50 @@ public void process(ProcessContext c) {
+  /** Wraps a {@link DataSourceConfiguration} to provide a {@link PoolingDataSource}. */
+  public static class PoolableDataSourceProvider
       implements SerializableFunction<Void, DataSource>, HasDisplayData {
-    private final DataSourceConfiguration config;
-    private static DataSourceProviderFromDataSourceConfiguration instance;
+    private static PoolableDataSourceProvider instance;
+    private static transient DataSource source;
+    private static SerializableFunction<Void, DataSource> dataSourceProviderFn;
 
-    private DataSourceProviderFromDataSourceConfiguration(DataSourceConfiguration config) {
-      this.config = config;
+    private PoolableDataSourceProvider(DataSourceConfiguration config) {
+      dataSourceProviderFn = DataSourceProviderFromDataSourceConfiguration.of(config);
     }
 
-    public static SerializableFunction<Void, DataSource> of(DataSourceConfiguration config) {
+    public static synchronized SerializableFunction<Void, DataSource> of(
+        DataSourceConfiguration config) {
       if (instance == null) {
-        instance = new DataSourceProviderFromDataSourceConfiguration(config);
+        instance = new PoolableDataSourceProvider(config);
       }
       return instance;
     }
 
     @Override
     public DataSource apply(Void input) {
-      return config.buildDatasource();
-    }
-
-    @Override
-    public void populateDisplayData(DisplayData.Builder builder) {
-      config.populateDisplayData(builder);
-    }
-  }
-
-  private abstract static class BaseDataSourceProvider
-      implements SerializableFunction<Void, DataSource>, HasDisplayData {
-    private final SerializableFunction<Void, DataSource> dataSourceProviderFn;
-
-    BaseDataSourceProvider(SerializableFunction<Void, DataSource> dataSourceProviderFn) {
-      this.dataSourceProviderFn = dataSourceProviderFn;
+      return buildDataSource(input);
+    }
+
+    static synchronized DataSource buildDataSource(Void input) {
+      if (source == null) {
+        DataSource basicSource = dataSourceProviderFn.apply(input);
+        DataSourceConnectionFactory connectionFactory =
+            new DataSourceConnectionFactory(basicSource);
+        PoolableConnectionFactory poolableConnectionFactory =
+            new PoolableConnectionFactory(connectionFactory, null);
+        GenericObjectPoolConfig poolConfig = new GenericObjectPoolConfig();
+        poolConfig.setMaxTotal(1);
+        poolConfig.setMinIdle(0);
+        poolConfig.setMinEvictableIdleTimeMillis(10000);
+        poolConfig.setSoftMinEvictableIdleTimeMillis(30000);
+        GenericObjectPool connectionPool =
+            new GenericObjectPool(poolableConnectionFactory, poolConfig);
+        poolableConnectionFactory.setPool(connectionPool);
+        poolableConnectionFactory.setDefaultAutoCommit(false);
+        poolableConnectionFactory.setDefaultReadOnly(false);
+        source = new PoolingDataSource(connectionPool);
+      }
+      return source;
```

Comment:
If `source` is provided but not a `PoolableDataSource`, we just use it, right (we don't wrap as a poolable datasource) ?